great_expectations/cli/datasource.py

```diff
@@ -447,15 +452,181 @@ def _add_spark_datasource(context, prompt_for_datasource_name=True):
+
+    return data_source
+
+def select_generator(context, data_source_name, available_data_assets_dict=None):
+    msg_prompt_select_generator = "Select generator"
+
+    if available_data_assets_dict is None:
+        available_data_assets_dict = context.get_available_data_asset_names(datasource_names=data_source_name)
+
+    available_data_asset_names_by_generator = {}
+    for key, value in available_data_assets_dict[data_source_name].items():
+        if len(value["names"]) > 0:
+            available_data_asset_names_by_generator[key] = value["names"]
+
+    if len(available_data_asset_names_by_generator.keys()) == 0:
+        return None
+    elif len(available_data_asset_names_by_generator.keys()) == 1:
+        return list(available_data_asset_names_by_generator.keys())[0]
+    else:  # multiple generators
+        generator_names = list(available_data_asset_names_by_generator.keys())
+        choices = "\n".join(["    {}. {}".format(i, generator_name) for i, generator_name in enumerate(generator_names, 1)])
+        option_selection = click.prompt(
+            msg_prompt_select_generator + "\n" + choices,
+            type=click.Choice([str(i) for i, generator_name in enumerate(generator_names, 1)]),
+            show_choices=False
+        )
+        generator_name = generator_names[int(option_selection)-1]
+
+        return generator_name
+
+def get_batch_kwargs(context,
+                  data_source_name=None,
+                  generator_name=None,
+                  data_asset_name=None,
+                  additional_batch_kwargs=None):
+    """
+    This method manages the interaction with user necessary to obtain batch_kwargs for a batch of a data asset.
+
+    In order to get batch_kwargs this method needs data_source_name, generator_name and data_asset_name
+    to combine them into a fully qualified data asset identifier(data_source_name/generator_name/data_asset_name).
+    All three arguments are optional. If they are present, the method uses their values. Otherwise, the method
+    prompts user to enter them interactively. Since it is possible for any of these three components to be
+    passed to this method as empty values and to get their values after interacting with user, this method
+    returns these components' values in case they changed.
+
+    If the datasource has generators that can list available data asset names, the method lets user choose a name
+    from that list (note: if there are multiple generators, user has to choose one first). If a name known to
+    the chosen generator is selected, the generator will be able to yield batch_kwargs. The method also gives user
+    an alternative to selecting the data asset name from the generator's list - user can type in a name for their
+    data asset. In this case the in-memory generator will be used to construct a fully qualified data asset
+    identifier (note: if the datasource has no in-memory generator configured, the method will exist with a failure).
```

Comment:
confirm this--does it *have* to be in-memory or can it be any of inmemory, passthrough, manual? (to *exit* vs *exist*)

Reply:
I'll update the doc to mention the passthrough generator. The manual will not work, because data assets have to be configured in it before they are referenced.